### ISSUE TITLE
Definition of "rot" suboptomal?

### ISSUE THREAD

**Issue by chdoc** [math-comp/math-comp#543]
I tried to use list (cycle) rotation today, I and I was surprised to see that `rot n s` only performs a rotation if `n <= size s`. In fact, there is even a lemma stating this. Moreover, several lemmas need (unexpected) side conditions:

```coq
rot_oversize: forall (T : Type) (n : nat) (s : seq T), size s <= n -> rot n s = s
rotD: forall (T : Type) (m n : nat) (s : seq T), m + n <= size s -> rot (m + n) s = rot m (rot n s)
rotS: forall (T : Type) (n : nat) (s : seq T), n < size s -> rot n.+1 s = rot 1 (rot n s)
rot_add_mod: forall (T : Type) (m n : nat) (s : seq T),
  n <= size s -> m <= size s -> rot m (rot n s) = rot (if m + n <= size s then m + n else m + n - size s) s
```
Wouldn't it be more natural to define rot as follows:
```coq
Definition rot n s :=  drop (n %% size s) s ++ take (n %% size s) s.
```
(the original definition is without the `_ %% size s`)

From this one can prove 
```coq
Lemma rot_size_r n s : rot (n + size s) s = rot n s.
Lemma rot_cons n x s : rot n.+1 (x::s) = rot n (rcons s x).
```
This allows removing the side conditions from `rotD` and `rotS`, making `rot_add_mod` redundant.

**Comment by CohenCyril** [comment]
I'm all in favor of this

**Comment by chdoc** [comment]
When I tried to replace the definition of `rot` I realized that currently `div` depends on `seq`. In particular `_ %% _` is not available in `seq.v`. So one would likely have to move `rot`, `rotr` and the associated lemmas from `seq.v` to another file that is further up the hierarchy. I guess `path.v` suggests itself, because that's where `rot` is used to rotate cycles.

**Comment by chdoc** [comment]
Yesterday, I tried to move the rotation lemmas to `path.v`, but it turns out that rot is also used in various lemmas in `seq.v` that do not mention `rot` in the statement. So it's not entirely clear to me what's the best thing to do here.

**Comment by chdoc** [comment]
To (maybe) make this progress eventually: I gave it another go using a definition of `rot` that doesn't rely on `_ %% _`:
```coq
Definition rot1 (T : eqType) (s : seq T) := if s is x::s' then rcons s' x else s.
Definition new_rot n (T : eqType) (s : seq T) := iter n (@rot1 _) s.
```
This leads to a reasonably nice theory. The key lemma is:
```coq
Lemma rot_small n s : n <= size s -> rot n s = drop n s ++ take n s.
```
combined with strengthening `rot_to` to not lose the information that one rotates at most `size _` times, this allows preserving the structure of most existing arguments (in `seq.v` at least). 

However, the new rotation function is noticeably slower, in particular on long lists:
```coq
Time Eval cbv in (old_rot 90 (nseq 100 0)). (* 0.002 - 0.004s *)
Time Eval cbv in (new_rot 90 (nseq 100 0)). (* 0.006 - 0.009s *)

Time Eval cbv in (old_rot 900 (nseq 1000 0)). (* 0.015 - 0.042s *)
Time Eval cbv in (new_rot 900 (nseq 1000 0)). (* 0.44  - 0.47s *)
```
This is a possible regression, but it would be easy to also provide 
```coq
Definition fast_rot n s :=  drop (n %% size s) s ++ take (n %% size s) s.
```
which is equivalent and for which the slowdown would be barely noticeable. 

Comments?

**Comment by chdoc** [comment]
> This leads to a reasonably nice theory.

I think I have to backtrack on this. I found [some proof](https://github.com/math-comp/math-comp/blob/e725070f9809a7d63f063a359dd7ef6bd3800dfd/mathcomp/ssreflect/seq.v#L3355) that repeatedly makes use of the fact that `rot n s` is of the shape `drop n' s ++ take n' s` (e.g., by rewriting with `rev (rot n _)` with `rev_cat`. While the proof can be fixed by inserting some uses of `rot_small`, this requires proving multiple side conditions that weren't necessary to prove before. 

In conclusion, I feel that the definition based on `iter` is a bad idea, in particular if this proof turns out to be representative of other proofs (e.g. in fourcolor). 

On the other hand, the definition I suggested initially:
```coq
Definition rot n s :=  drop (n %% size s) s ++ take (n %% size s) s.
```
would cause less breakage. 

As it relies on `modn`, this would require some reordering in the library though. This is clearly possible, since the theory of `rot` would only need some basic facts about `modn`, and the first part of `div.v` that actually uses `seq` is the function computing Bezout coefficients. One option would be to move necessary parts (say everything up to the definition of `gcdn`) of `div.v` to `ssrnat.v`.

**Comment by chdoc** [comment]
I tried using the definition based on `_ %% _`, but that still leads to breakage at various places, some of them not entirely trivial to fix. Maybe it is better to keep the current definition and add some additional lemmas that ease reasoning about nested applications of `rot` along the lines of the following:
```coq
Section RotAddMod.
Variables (T : eqType).
Implicit Type (s : seq T).

Lemma rot_inord n s : rot n s = rot (inord n : 'I_(size s).+1) s.
Proof. 
rewrite /inord/insubd; case: insubP => //= [u ? -> //|].
by rewrite -leqNgt => ?; rewrite rot0 rot_oversize // ltnW.
Qed.

Lemma rot_combine n m s : { o : nat | o <= size s & rot m (rot n s) = rot o s }.
Proof.
eexists; last first.
  by rewrite (rot_inord n) (rot_inord m) rot_add_mod // -ltnS ?size_rot ltn_ord.
by case:ifP => // _; rewrite -subnBA // ?leq_subn ?size_rot // -ltnS ltn_ord.
Qed.

Definition rot_sum n m s := sval (@rot_combine n m s).

Lemma leq_rot_sum n m s : rot_sum n m s <= size s.
Proof. exact: (s2valP (rot_combine n m s)). Qed.

Lemma rot_rot_sum n m s : rot m (rot n s) = rot (rot_sum n m s) s.
Proof. exact: (s2valP' (rot_combine n m s)). Qed.
End RotAddMod.
```
I'd be glad for suggestions, in particular for better names. Using `inord` is maybe not ideal, because the the hidden implicit argument can also be something like `size (rot m s)`. So maybe this "rotation normalization" should be another definition.

**Comment by CohenCyril** [comment]
@chdoc unless I am mistaken, this was fixed in  #603?

**Comment by chdoc** [comment]
I'm not sure I'd use the word "fixed" since #603 still keeps the truncating rotation I was trying to get rid of. But you're right, I gave up on changing the definition of `rot`. For the record, this is because there quite a few proofs that rely on the definition expanding to `drop n s ++ take n s` without having to prove that `n <= size s`. Any change in the definition would require proving side conditions in these cases. PR #603 makes reasoning about compositions of `rot` bearable, so this issue can be closed.
